contributors/design-proposals/apps/configmap-secret-gc.md

```diff
@@ -0,0 +1,271 @@
+
+
+### Embed ConfigMaps/Secrets in Controller Spec
+
+Add `.configMapTemplate` and `.secretTemplate` fields to a controller's `.spec`. 
+
+**Cons:**
+
+*   Adds complexity to the API, tooling dealing with ConfigMaps/Secrets, and the
+    implementation of controllers.
+*   Controller spec becomes fat.
+
+
+### Specify ConfigMaps/Secrets Garbage Collection in Controller Spec
+
+Add a `bool` field in controller `.spec` to specify whether garbage collection
+of its ConfigMaps/Secrets is enabled. For more granularity, add another field
+allowing users to specify a list of ConfigMaps/Secrets that should not be
+garbage collected even if the aforementioned field is enabled. 
+
+**Cons:**
+
+*   User intent conflict. For example, user A wants a ConfigMap to be garbage
+    collected with a Deployment, and therefore the ConfigMap will be deleted
+    after the Deployment is gone. User B cannot stop the ConfigMap from being
+    deleted even if B disables garbage collection of that ConfigMap in B's
+    controller spec. 
+*   To enable or disable garbage collection of a ConfigMap/Secret, users need to
+    update spec of all controllers that reference it. 
+
+
+### Specify ConfigMaps/Secrets garbage collection in PodSpec
+
+Add a `bool` field everywhere in `PodSpec` that ConfigMaps/Secrets are
+referenced (e.g. `.containers.envFrom.configMapRef`) to specify whether garbage
+collection of the ConfigMap/Secret is enabled. 
+
+**Cons:**
+
+*   Verbosity. if a ConfigMap is referenced in multiple places in a pod template,
+    users need to specify it everywhere. 
+*   User intent conflict.
+*   To enable or disable garbage collection of a ConfigMap/Secret, users need to
+    update pod template of all controllers that reference it. For controllers
+    with history, this means triggering rollouts.
+
+
+### Automatic ConfigMap/Secret References Updates
+
+Every time a ConfigMap/Secret is updated, the controller automatically creates a
+new ConfigMap/Secret (the new name can be generated by appending a hash), and
+updates references in pod templates in controller history objects. Every
+ConfigMap/Secret update triggers a rollout.
+
+For example, make Deployment controller create an ConfigMap "abc-<hash>" from
+ConfigMap "abc" referenced in its pod template, and then reference "abc-<hash>"
+in ReplicaSet pod template. Or something like
+[#31701](https://github.com/kubernetes/kubernetes/pull/31701).
+
+This is not strictly about garbage collection of ConfigMaps/Secrets, but is
+relevant to the garbage collection design. 
+
+**Pros:**
+
+*   ConfigMap/Secret references are updated automatically.
+
+**Cons:**
+
+*   It becomes tricky for controllers (e.g. Deployments) to compare its pod
+    template against its history (e.g. ReplicaSets) to find current history
+    (ReplicaSet). 
+*   Rollbacks trigger updates to original ConfigMaps. All consumers of a
+    ConfigMap/Secret are then forced to do rollbacks together. 
```

Comment:
I think this section is the right approach - lift the responsibility for cloning CM/S up into a top level controller and have it selectively mutate top level controllers as requested.  Define a linkage from a controller object back to a cm template, and make the new controller able to work with many different API types (possibly by using unstructured json path on annotations).

Reply:
One of the goals of CM rollout is that it should become possible to roll back a Deployment, and have it revert to the exact, known-good combination of CM revision and Pod Template.

Once a Deployment is rolled back, how will the top-level "CM update" controller know it shouldn't bump it back to the newest CM revision?

Reply:
You need to turn off CM update if you want to roll back.  You were going to have to do that anyway, because the controller for cm update could bump you at any time resulting in a new deployment.  If you want to roll the state back to a previous version, you either revert the config (in which case this approach works the same as before) or you disable trigger of the configmap and then rollback.

Another advantage of this approach - the deployment is exactly exportable (to a known good state) whereas in the other approach the user would have to export the deployment and the replica set.

Reply:
The benefit of the model that @janetkuo proposes is that the entire configuration is encapsulated in the config map itself. For any of the controllers, if you point them to a previous config map or secret, they will detect the identical previous state, and perform a minimally disruptive rollback. If you export the Deployment, with a reference to the config map, and reapply it, you will get the same behavior. Also, both approaches can be used in conjunction with kubectl rollback. I think the proposed approach is actually more flexible.